stablehlo/reference/Ops.cpp

```diff
@@ -133,6 +322,182 @@ Tensor evalConvertOp(const Tensor &operand, TensorType resultType) {
+    auto rhses = splitIntoNGroupsAlongDim(rhs, featureGroupCount,
+                                          kernelOutputFeatureDimension);
+    SmallVector<Tensor> results;
+    for (auto [left, right] : llvm::zip(lhses, rhses)) {
+      auto convolutionResult = evalConvolutionOp(
+          left, right, windowStrides, padding, lhsDilation, rhsDilation,
+          windowReversal, inputBatchDimension, inputFeatureDimension,
+          inputSpatialDimensions, kernelInputFeatureDimension,
+          kernelOutputFeatureDimension, kernelSpatialDimensions,
+          outputBatchDimension, outputFeatureDimension, outputSpatialDimensions,
+          /*featureGroupCount=*/1, batchGroupCount,
+          inferConvolutionOpType(
+              left.getType(), right.getType(), windowStrides, padding,
+              lhsDilation, rhsDilation, windowReversal, inputBatchDimension,
+              inputFeatureDimension, inputSpatialDimensions,
+              kernelInputFeatureDimension, kernelOutputFeatureDimension,
+              kernelSpatialDimensions, outputBatchDimension,
+              outputFeatureDimension, outputSpatialDimensions,
+              /*featureGroupCount=*/1, batchGroupCount,
+              /*precisionConfig=*/{}, resultType));
+      results.push_back(convolutionResult);
+    }
+    return evalConcatenateOp(results, outputFeatureDimension, result.getType());
+  }
+  if (batchGroupCount > 1) {
+    auto lhses =
+        splitIntoNGroupsAlongDim(lhs, batchGroupCount, inputBatchDimension);
+    auto rhses = splitIntoNGroupsAlongDim(rhs, batchGroupCount,
+                                          kernelOutputFeatureDimension);
+    SmallVector<Tensor> results;
+    for (auto [left, right] : llvm::zip(lhses, rhses)) {
+      auto convolutionResult = evalConvolutionOp(
+          left, right, windowStrides, padding, lhsDilation, rhsDilation,
+          windowReversal, inputBatchDimension, inputFeatureDimension,
+          inputSpatialDimensions, kernelInputFeatureDimension,
+          kernelOutputFeatureDimension, kernelSpatialDimensions,
+          outputBatchDimension, outputFeatureDimension, outputSpatialDimensions,
+          featureGroupCount, /*batchGroupCount=*/1,
+          inferConvolutionOpType(
+              left.getType(), right.getType(), windowStrides, padding,
+              lhsDilation, rhsDilation, windowReversal, inputBatchDimension,
+              inputFeatureDimension, inputSpatialDimensions,
+              kernelInputFeatureDimension, kernelOutputFeatureDimension,
+              kernelSpatialDimensions, outputBatchDimension,
+              outputFeatureDimension, outputSpatialDimensions,
+              featureGroupCount, /*batchGroupCount=*/1,
+              /*precisionConfig=*/{}, resultType));
+      results.push_back(convolutionResult);
+    }
+    return evalConcatenateOp(results, outputFeatureDimension, result.getType());
+  }
+  auto lhsWindowDimensions = concatAndPermute(
```

Comment:
> auto lhsPermutation = inputBatchDimension + inputSpatialDimensions + inputFeatureDimension;

Since we can't quite use the syntax sugar mentioned above, I moved the computation for permutation out of the helper function. This saves some compute and is also closer to the spec.

Reply:
Yes, for that we need to added overload `operator+`  in Axes? Let us weigh in @burmako opinion on this. Keeping this unresolved for now.